pkg/storage/crd/storage_interface.go

```diff
@@ -0,0 +1,547 @@
+	scmeta "github.com/kubernetes-incubator/service-catalog/pkg/api/meta"
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/apiserver/pkg/storage/etcd"
+	"k8s.io/apiserver/pkg/storage/storagebackend/factory"
+	restclient "k8s.io/client-go/rest"
+)
+
+var (
+	errNotImplemented = errors.New("not implemented for third party resources")
+)
+
+type store struct {
+	// TODO (nilebox): review the usage of "hasNamespace" flag
+	hasNamespace     bool
+	defaultNamespace string
+	cl               restclient.Interface
+	resourcePlural   string
+	// singularShell is a function that returns a new object of the appropriate type,
+	// with the namespace (first param) and name (second param) pre-filled
+	singularShell func(string, string) runtime.Object
+	listKind      Kind
+	// listShell is a function that returns a new, empty list object of the appropriate
+	// type. The list object should hold elements that are returned by singularShell
+	listShell   func() runtime.Object
+	checkObject func(runtime.Object) error
+	decodeKey   func(string) (string, string, error)
+	versioner   storage.Versioner
+	hardDelete  bool
+}
+
+// NewStorage creates a new TPR-based storage.Interface implementation
+func NewStorage(opts Options) (storage.Interface, factory.DestroyFunc) {
+	return &store{
+		hasNamespace:     opts.HasNamespace,
+		defaultNamespace: opts.DefaultNamespace,
+		cl:               opts.RESTClient,
+		resourcePlural:   opts.ResourcePlural,
+		singularShell:    opts.NewSingularFunc,
+		listKind:         opts.ListKind,
+		listShell:        opts.NewListFunc,
+		checkObject:      opts.CheckObjectFunc,
+		decodeKey:        opts.Keyer.NamespaceAndNameFromKey,
+		versioner:        etcd.APIObjectVersioner{},
+		hardDelete:       opts.HardDelete,
+	}, opts.DestroyFunc
+}
+
+// Versioned returns the versioned associated with this interface
+func (t *store) Versioner() storage.Versioner {
+	return t.versioner
+}
+
+// Create adds a new object at a key unless it already exists. 'ttl' is time-to-live
+// in seconds (0 means forever). If no error is returned and out is not nil, out will be
+// set to the read value from database.
+func (t *store) Create(
+	ctx context.Context,
+	key string,
+	obj,
+	out runtime.Object,
+	ttl uint64,
+) error {
+	// TODO (nilebox): can't we just use meta.Accessor(obj) instead?
+	ns, name, err := t.decodeKey(key)
+	if err != nil {
+		glog.Errorf("decoding key %s (%s)", key, err)
+		return err
+	}
+
+	if err := scmeta.AddFinalizer(obj, v1alpha1.FinalizerServiceCatalog); err != nil {
+		glog.Errorf("adding finalizer to %s (%s)", key, err)
+		return err
+	}
+
+	req := t.cl.Post().Resource(t.resourcePlural).Namespace(ns).Name(name).Body(obj)
```

Comment:
I believe that setting up a REST client with the scheme supporting all Service Catalog kinds, this approach should work